pvlib/mlfm.py

```diff
@@ -0,0 +1,804 @@
+    gap = 0.01
+
+    inv_ff = 1 / fill_factor
+
+    if all(c in dnorm.columns for c in ['v_ff', 'r_oc', 'i_ff', 'r_sc']):
+
+        # include effects of series and shunt resistances in stacked losses
+        # find factor to transform multiplicative to subtractive losses
+        # correction factor to scale losses to keep 1/ff --> pr_dc
+
+        # product
+        prod = inv_ff * (
+            dnorm['i_sc'] * dnorm['r_sc'] * dnorm['i_ff'] *
+            dnorm['v_ff'] * dnorm['r_oc'] * dnorm['v_oc']
+        )
+
+        # total
+        tot = inv_ff + (
+            dnorm['i_sc'] + dnorm['r_sc'] + dnorm['i_ff'] +
+            dnorm['v_ff'] + dnorm['r_oc'] + dnorm['v_oc'] - 6
+        )
+
+        # correction factor
+        corr = (inv_ff - prod) / (inv_ff - tot)
+
+        # put mlfm values in a stack from pr_dc (bottom) to 1/ff_ref (top)
+        # accounting for series and shunt resistance losses
+        dstack['pr_dc'] = +dnorm['pr_dc']  # initialise
+        dstack['i_sc'] = -(dnorm['i_sc'] - 1) * corr
+        dstack['r_sc'] = -(dnorm['r_sc'] - 1) * corr
+        dstack['i_ff'] = -(dnorm['i_ff'] - 1) * corr - gap/2
+        dstack['i_v'] = gap
+        dstack['v_ff'] = -(dnorm['v_ff'] - 1) * corr - gap/2
+        dstack['r_oc'] = -(dnorm['r_oc'] - 1) * corr
+        dstack['v_oc'] = -(dnorm['v_oc'] - 1) * corr
+        dstack['temp_module_corr'] = (
+            -(dnorm['v_oc'] - dnorm['v_oc_temp_corr']) * corr)
+
+        return dstack
+
+    # subtractive losses without series and shunt resistance effects
+    # find factor to transform multiplicative to subtractive losses
+    # correction factor to scale losses to keep 1/ff --> pr_dc
+
+    prod = inv_ff * (
+        dnorm['i_sc'] * dnorm['i_mp'] *
+        dnorm['v_mp'] * dnorm['v_oc']
+    )
+
+    tot = inv_ff + (
+        dnorm['i_sc'] + dnorm['i_mp'] +
+        dnorm['v_mp'] + dnorm['v_oc'] - 4
+    )
+
+    corr = (inv_ff - prod) / (inv_ff - tot)
+
+    # put mlfm values in a stack from pr_dc (bottom) to 1/ff_ref (top)
+    dstack['pr_dc'] = + dnorm['pr_dc']  # initialise
+    dstack['i_sc'] = -(dnorm['i_sc'] - 1) * corr
+    dstack['i_mp'] = -(dnorm['i_mp'] - 1) * corr - gap/2
+    dstack['i_v'] = gap
+    dstack['v_mp'] = -(dnorm['v_mp'] - 1) * corr - gap/2
+    dstack['v_oc'] = -(dnorm['v_oc'] - 1) * corr
+
+    dstack['temp_module_corr'] = (
+        - (dnorm['v_oc'] - dnorm['v_oc_temp_corr']) * corr)
+
+    return dstack
+
+
+def mlfm_6(dmeas, c_1, c_2, c_3, c_4, c_5=0., c_6=0.):
+    r'''
+    Predict normalised LFM values from data in ``dmeas``.
+
+    The normalized LFM values are given by
+
+    .. math::
+
+        c_1 + c_2 (T_m - 25) + c_3 \log10(G_{POA}) + c_4 G_{POA}
+        + c_5 WS + c_6 / G_{POA}
+
+    where :math:`G_{POA}` is global plane-of-array (POA) irradiance in kW/m2,
+    :math:`T_m` is module temperature in C and :math:`WS` is wind speed in
+    m/s.
+
+    Parameters
+    ----------
+    dmeas : DataFrame
```

Comment:
Suggest making three separate parameters for the three input columns as is usually done in pvlib.